reference-implementation/lib/writable-stream.js

```diff
@@ -234,65 +672,109 @@ function IsWritableStream(x) {
   return true;
 }
 
-exports.IsWritableStream = IsWritableStream;
+function WritableStreamDefaultControllerAdvanceQueueIfNeeded(controller) {
+  const state = controller._controlledWritableStream._state;
+  if (state === 'closed' || state === 'errored') {
+    return;
+  }
 
-function SyncWritableStreamStateWithQueue(stream) {
-  if (stream._state === 'closing') {
-    return undefined;
+  if (controller._started === false) {
+    return;
   }
 
-  assert(stream._state === 'writable' || stream._state === 'waiting',
-    'stream must be in a writable or waiting state while calling SyncWritableStreamStateWithQueue');
+  if (controller._queue.length === 0) {
+    return;
+  }
 
-  const queueSize = GetTotalQueueSize(stream._queue);
-  const shouldApplyBackpressure = queueSize > stream._strategyHWM;
+  if (controller._writing === true) {
+    return;
+  }
 
-  if (shouldApplyBackpressure === true && stream._state === 'writable') {
-    stream._state = 'waiting';
-    stream._readyPromise = new Promise((resolve, reject) => {
-      stream._readyPromise_resolve = resolve;
-    });
+  const writeRecord = PeekQueueValue(controller._queue);
+  if (writeRecord === 'close') {
+    WritableStreamDefaultControllerProcessClose(controller);
+  } else {
+    WritableStreamDefaultControllerProcessWrite(controller, writeRecord.chunk);
   }
+}
 
-  if (shouldApplyBackpressure === false && stream._state === 'waiting') {
-    stream._state = 'writable';
-    stream._readyPromise_resolve(undefined);
+function WritableStreamDefaultControllerErrorIfNeeded(controller, e) {
+  const state = controller._controlledWritableStream._state;
+  if (state === 'writable' || state === 'waiting' || state === 'closing') {
+    WritableStreamDefaultControllerError(controller, e);
   }
+}
+
+function WritableStreamDefaultControllerProcessClose(controller) {
+  const stream = controller._controlledWritableStream;
+
+  assert(stream._state === 'closing', 'can\'t process final write record unless already closed');
+
+  DequeueValue(controller._queue);
+  assert(controller._queue.length === 0, 'queue must be empty once the final write record is dequeued');
+
+  const sinkClosePromise = PromiseInvokeOrNoop(controller._underlyingSink, 'close');
+  sinkClosePromise.then(
+    () => {
+      if (stream._state !== 'closing') {
+        return;
+      }
 
-  return undefined;
+      WritableStreamFulfillWriteRequest(stream);
```

Comment:
_writeRequests now holds both pending write and pending close. That's why we're calling this here.